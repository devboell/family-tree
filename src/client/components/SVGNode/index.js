import React from 'react'
import { PropTypes } from 'prop-types'

import SVGRelation from 'components/SVGRelation'
import SVGRelationLink from 'components/SVGRelationLink'

import PersonInfoPanel from 'components/PersonInfoPanel'


const SVGNode = ({
  node,
  offsetX,
  linkTo,
  boxWidth,
  marginX,
  boxHeight,
  marginY,
}) => {
  // if a  node has relation(s), it is placed to the left of the center
  // of the subtree of the first relation.
  const subTreeWidths = node.partners.map(relation => relation.subTreeWidth)
  const x = subTreeWidths.length
    ? offsetX + ((subTreeWidths[0] / 2) - boxWidth)
    : offsetX + marginX


  const y = node.level * (boxHeight + marginY)
  const midPointOffset = arr =>
    // eslint-disable-next-line no-confusing-arrow
    arr.reduce((acc, val, i) =>
    i === arr.length - 1
      ? acc + (val / 2)
      : acc + val
    , 0)


  const midPoint = i => midPointOffset(subTreeWidths.slice(0, i + 1))
  const prevMidPoint = i => midPointOffset(subTreeWidths.slice(0, i))

  const target = i => (
    i > 0
    ? offsetX + prevMidPoint(i) + marginX + boxWidth
    : (offsetX + (subTreeWidths[0] / 2))
  )

  const prevRelOffset = i => node.partners.slice(0, i)
    .reduce((acc, rel) => acc + rel.subTreeWidth, 0)

  return (
    <g style={{ shapeRendering: 'crispEdges' }} stroke="black" strokeWidth="1" fill="none">
      <foreignObject
        {...{ x }}
        {...{ y }}
        width={boxWidth}
        height={boxHeight}
      >
        <PersonInfoPanel
          width={boxWidth}
          height={boxHeight}
        >
          <p>{node.value}</p>
        </PersonInfoPanel>
      </foreignObject>
      {(linkTo)
        ? <path d={`M ${x + (boxWidth / 2)} ${y} V ${y - (marginY / 2)} H ${linkTo.x} V ${linkTo.y}`} />
        : null
      }
      {node.partners.map((relation, i) => (
        <g key={`relation-${relation.value}`}>
          <SVGRelationLink
            originX={offsetX + midPoint(i) + marginX}
            originY={y + (boxHeight / 2)}
            targetX={target(i) + 2}
          />
          <SVGRelation
            relationNode={relation}
            {...{ y }}
            offsetX={offsetX + prevRelOffset(i)}
            {...{ boxWidth, marginX, boxHeight, marginY }}
          />
        </g>
      ))}
    </g>
  )
}

SVGNode.propTypes = {
  node: PropTypes.object.isRequired,
  offsetX: PropTypes.number.isRequired,
  linkTo: PropTypes.object,
  boxWidth: PropTypes.number.isRequired,
  marginX: PropTypes.number.isRequired,
  boxHeight: PropTypes.number.isRequired,
  marginY: PropTypes.number.isRequired,
}


export default SVGNode
