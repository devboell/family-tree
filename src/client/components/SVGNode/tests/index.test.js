import React from 'react'
import SVGNode from '../index'

const withPartnerNode = {
  level: 0,
  partners: [
    {
      children: [
        {
          level: 1,
          partners: [],
          totalHeight: 100,
          totalWidth: 100,
          value: 'child 1 - level 1',
        },
        {
          level: 1,
          partners: [],
          totalHeight: 100,
          totalWidth: 100,
          value: 'child 2 - level 1',
        },
      ],
      level: 0,
      totalHeight: 200,
      totalWidth: 200,
      value: 'root - partner 1',
    },
    {
      children: [],
      level: 0,
      totalHeight: 100,
      totalWidth: 200,
      value: 'root - partner 2',
    },
  ],
  totalHeight: 200,
  totalWidth: 400,
  value: 'root',
}

const withoutPartnerNode = {
  level: 0,
  partners: [],
  totalHeight: 100,
  totalWidth: 100,
  value: 'root',

}
const props = {
  node: {},
  offsetX: 200,
  linkTo: null,
  dimensions: {
    boxWidth: 100,
    boxHeight: 100,
    marginX: 50,
    marginY: 50,
  },
}

describe('SVGNode component', () => {
  it('SVGNode, renders node with partners', () => {
    const newProps = {
      ...props,
      node: withPartnerNode,
    }
    const wrapper = shallow(<SVGNode {...newProps} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('SVGNode, renders node without partners', () => {
    const newProps = {
      ...props,
      node: withoutPartnerNode,
    }
    const wrapper = shallow(<SVGNode {...newProps} />)
    expect(wrapper).toMatchSnapshot()
  })
})
