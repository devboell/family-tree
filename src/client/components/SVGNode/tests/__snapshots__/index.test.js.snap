// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SVGNode component SVGNode, renders node with partners 1`] = `
<g
  fill="none"
  stroke="black"
  strokeWidth="1"
  style={
    Object {
      "shapeRendering": "crispEdges",
    }
  }
>
  <SVGNodeInfo
    boxHeight={100}
    boxWidth={100}
    info="root"
    x={200}
    y={0}
  />
  <SVGNodeLink
    dimensions={
      Object {
        "boxHeight": 100,
        "boxWidth": 100,
        "marginX": 50,
        "marginY": 50,
      }
    }
    linkTo={null}
    x={200}
    y={0}
  />
  <SVGPartnerNode
    dimensions={
      Object {
        "boxHeight": 100,
        "boxWidth": 100,
        "marginX": 50,
        "marginY": 50,
      }
    }
    node={
      Object {
        "children": Array [
          Object {
            "level": 1,
            "partners": Array [],
            "totalHeight": 100,
            "totalWidth": 100,
            "value": "child 1 - level 1",
          },
          Object {
            "level": 1,
            "partners": Array [],
            "totalHeight": 100,
            "totalWidth": 100,
            "value": "child 2 - level 1",
          },
        ],
        "level": 0,
        "totalHeight": 200,
        "totalWidth": 200,
        "value": "root - partner 1",
      }
    }
    offsetX={200}
    previousSubTreeWidth={0}
    y={0}
  />
  <SVGPartnerNode
    dimensions={
      Object {
        "boxHeight": 100,
        "boxWidth": 100,
        "marginX": 50,
        "marginY": 50,
      }
    }
    node={
      Object {
        "children": Array [],
        "level": 0,
        "totalHeight": 100,
        "totalWidth": 200,
        "value": "root - partner 2",
      }
    }
    offsetX={400}
    previousSubTreeWidth={200}
    y={0}
  />
</g>
`;

exports[`SVGNode component SVGNode, renders node without partners 1`] = `
<g
  fill="none"
  stroke="black"
  strokeWidth="1"
  style={
    Object {
      "shapeRendering": "crispEdges",
    }
  }
>
  <SVGNodeInfo
    boxHeight={100}
    boxWidth={100}
    info="root"
    x={250}
    y={0}
  />
  <SVGNodeLink
    dimensions={
      Object {
        "boxHeight": 100,
        "boxWidth": 100,
        "marginX": 50,
        "marginY": 50,
      }
    }
    linkTo={null}
    x={250}
    y={0}
  />
</g>
`;
